File: 783.Minimum-Distance-Between-BST-Nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function (root) {
  let diff = Infinity;
  let prev = null;

  const traverse = (node) => {
    if (node.left) traverse(node.left);

    if (prev) {
      diff = Math.min(diff, node.val - prev.val);
    }

    prev = node;

    if (node.right) traverse(node.right);
  };

  traverse(root);

  return diff;
};

File: python/783.Minimum-Distance-Between-BST-Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        prev = None
        diff = float('inf')

        def traverse(node):
            if not node:
                return

            traverse(node.left)
            
            nonlocal prev, diff

            if prev:
                diff = min(diff, node.val - prev.val)
            prev = node

            traverse(node.right)
        
        traverse(root)

        return diff